Document inferred const args (`feature(generic_arg_infer)`)

BoxyUwU opened the pull request:
Tracking: https://github.com/rust-lang/rust/issues/85077
Stabilization: https://github.com/rust-lang/rust/pull/141610
Stabilization report: https://github.com/rust-lang/rust/pull/141610#issuecomment-2910505867

BoxyUwU commented on src/const-generics.md:
````diff
@@ -0,0 +1,174 @@
+```
+
+r[const-generics.concrete-expr]
+## Concrete expressions
````
Something I left out here is that the "concrete expressions" are not required to be quite so concrete for repeat expr counts as a backwards compatibility FCW. Does the reference intend to document the *actual state* of the compiler, or can I take liberties in documenting things as if FCWs are already hard errors?

BoxyUwU commented:
Sure, done

BoxyUwU commented:
@rustbot ready

BoxyUwU commented:
@ehuss @traviscross any chance y'all will be able to get around to reviewing this soon? The stabilization PR is now ready to be merged and it'd be unfortunate to block a relatively minor feature that's already been waiting years, on this PR.

ehuss commented:
If possible, I think it would be best to drop the rearrangement commit. This probably isn't the organization we'll land on, and I would prefer to keep these as separate concerns so they don't block one another.

BoxyUwU committed:
Document inferred const args

BoxyUwU commented:
Ok this should be fine now

BoxyUwU commented on src/items/generics.md:
````diff
@@ -227,6 +227,22 @@ fn generic<const B: bool>() {
+```grammar,types
+InferredConst -> 
+      `_`
+    | (InferredConst)
````
okay @ root didn't fix anything. I'm not sure how i'm supposed to use this anywhere

BoxyUwU commented:
@traviscross @ehuss friendly poke about this :)

ehuss commented on src/items/generics.md:
````diff
@@ -227,6 +227,22 @@ fn generic<const B: bool>() {
+r[items.generics.const.inferred]
+
+r[items.generics.const.inferred.syntax]
+```grammar,const
+InferredConst ->
+      `_`
+    | `(` InferredConst `)`
+```
````
I'm not sure I fully understand this grammar since it isn't referred to anywhere.

Am I understanding correctly that this is added for GenericArgsConst? Or is it somewhere else? That is, should it be:

```
GenericArgsConst ->
      BlockExpression
    | LiteralExpression
    | `-` LiteralExpression
    | SimplePathSegment
    | InferredConst

InferredConst ->
      `_`
    | `(` InferredConst `)`
```

ehuss commented on src/items/generics.md:
```diff
@@ -227,6 +227,22 @@ fn generic<const B: bool>() {
+The inferred const asks the compiler to infer the const argument if possible
+based on the surrounding information available.
+
+r[items.generics.const.inferred.constraint]
+It cannot be used in item signatures.
```
I'm not quite following this. If we are defining this as a new form allowed in generic *args*, why would there be a rule that it is not allowed in item signatures?

BoxyUwU replied:
We don't allow type inference in signatures in general, not just as a restriction on const arguments. For example this is an error on stable:
```rust
struct Foo<T>(T);
fn bar(param: Foo<_>) {}
```

ehuss replied:
Isn't this already stated in type.inferred.constraint (inferred.md)?

BoxyUwU replied:
Yeah we state that for `_`  type arguments but I felt like like we should note that the restriction applies for `_` const arguments too when discussing them as it might be confusing otherwise; people might get the idea that you *can* use `_` const arguments in signatures.

ehuss commented on src/items/generics.md:
```diff
@@ -227,6 +227,22 @@ fn generic<const B: bool>() {
+
+r[items.generics.const.inferred.intro]
+The inferred const asks the compiler to infer the const argument if possible
+based on the surrounding information available.
```
Can you add a semi-realistic looking example here?

ehuss reviewed (commented):
I'm a little lost, as it seems like there are two distinct things being supported here:

1. `_` in generic args: `foo::<_>()`
2. `_` in array expression repeat lengths: `[123; _]`

Am I understanding that correctly? If so, shouldn't there be some update to the array expression chapter? And the underscore-expr.md may need a significant update to mention that it is allowed for array repeats.

BoxyUwU commented:
> I'm a little lost, as it seems like there are two distinct things being supported here

There is one thing being added, we support `_` (optionally wrapped in arbitrary many parens) as const arguments. The repeat count of an array expr is a const argument so falls under the same thing as in `let a: [u8; _] = ...`. This is equivalent in functionality to how we support `_` (optionally wrapped in arbitrary many parens) as type arguments (e.g. `let a: [_; 2]`).

I didn't think the array-expr chapter would need updating as I've updated the definition of a const argument but it seems like the reference attempts to separately specify repeat expr counts from const arguments. This also means that the chapter does not currently talk about the fact that `[0x1; N + 1]` is illegal. I'm not sure what to do about that, can we just leave it as-is or would you prefer me to update the array-expr chapter to specify repeat counts as being const arguments.

> And the underscore-expr.md may need a significant update

I don't think we need to update underscore-expr, this feature is (confusingly) entirely orthogonal from `_` as an expression. `[u8; _]` isnt giving the `_` expression new meaning, its adding a new *kind* of const argument that is not an expression.

BoxyUwU commented:
Sorry I feel like I'm repeating myself a bit here and am not sure how to explain this better. This feature is orthogonal to underscores as an expression. It's not that `_` is now a valid expression by itself, it's that there is a new kind of *const generic argument* (which happens to have the same syntax as the _ expression).

Maybe the core confusion here is about this distinction between a const *expression* and a const *argument*?

> If I'm understanding correctly, we define the array repeat length as a [const expression](https://doc.rust-lang.org/nightly/reference/const_eval.html#r-const-eval.const-expr.const-context)

I think the entire way that const generics are handled in the reference right now is just not structured right for this feature to be honest ^^.

It doesn't make sense to talk about const arguments as being "expressions with more restrictions"; `_` arguments are *not* expressions and are instead a new *kind* of const argument. But as far as I can tell there is no distinct concept of a "const argument" in the reference (like how there are different kinds of for type arguments). Instead everywhere simply specifies things as being "const expressions" with some additional restrictions.

I originally tried to refactor the reference to properly talk about const arguments as a real concept in the type system section with different sections for each kind of const argument. But that was quite a big change and this seemed smaller/easier to review.

Do you think you'd prefer going back to the old approach that has more involved explanations of this topic? It sounds like the new version of this PR is actually quite confusing in practice which is unfortunate.

Admittedly, I'm finding this PR to be quite stressful. I'd really like to get `generic_arg_infer` stabilized and having it be blocked on this when there's such a disconnect between the language itself and how the reference specifies things is making me worried that it's going to wind up blocked for a really long time.

I'd be more than happy to put a bunch of time into documenting things in the reference but if that's where things need to go in order to get something you're happy with would it be possible to just land the stabilization PR *first* and continue work on documenting this afterwards?

> I'm not quite following this. That looks like an _ in type position

I guess with this terminology it would be correct to say that the repeat count of array exprs is a type position ^^. Though being nitpicky it doesn't make sense to say that a `_` there is an inferred type because it is not a type, its a const generic value. Hence why I've added the concept of an "inferred const" to the reference which roughly mirrors the structure of the inferred type documentation.

> I believe these limitations are defined in https://doc.rust-lang.org/nightly/reference/const_eval.html#const-context and https://doc.rust-lang.org/nightly/reference/items/generics.html#r-items.generics.const.standalone.

Ah yes you're right it is documented :+1: Then I guess array-expr really doesn't need updating because it is specified as being a const generic argument :D

BoxyUwU commented on src/paths.md:
```diff
@@ -65,6 +65,7 @@ GenericArgsConst ->
     | LiteralExpression
     | `-` LiteralExpression
     | SimplePathSegment
+    | InferredConst
```
Possibly also a `ParenthesisedTyOrConst` thing on `GenericArgs` to solve the parens issue too

BoxyUwU replied:
Basically you'd split `GenericArgsConst/Type` into `GenericArgsConst/TypeUnambig/Ambig` and then lift the shared syntax out of the `Ambig` versions and onto `GenericArg` which would also be changed to use the `Ambig` forms

BoxyUwU commented on src/expressions/array-expr.md:
```diff
@@ -26,16 +26,16 @@ r[expr.array.array-behavior]
 
 r[expr.array.repeat]
-The syntax for the second form is two expressions separated by a semicolon (`;`) enclosed in square brackets.
+The syntax for the second form is an expression and a const argument separated by a semicolon (`;`) enclosed in square brackets.
```
as in, it consists of *both* an expression and a const argument. it is a bit ambiguous let me update that

BoxyUwU replied:
[expr; constargument]

traviscross committed:
Revise a bit editorially

BoxyUwU committed:
Dont add new grammar rules

traviscross committed:
Improve the `generic_arg_infer` docs

Syntactically, the inferred const is either an expression or an
`InferredType`, depending on the context.  Semantically, though, it's
a new kind of const generic argument.  Let's add the necessary
language in various places to make this all as clear as possible.
